### PR TITLE
Add Julia support

### PR DESCRIPTION
**Pull request by josephsdavid**
Slowly figuring out how to add julia support, currently just for functions and structs as that is all i know how to do, but eventually maybe macro docs... todos:

- [x] struct documentation (much easier since there aren't 10 million types of arguments)
- [x] function documentation

### CODE REVIEW

File: lua/neogen/configurations/julia.lua [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                        local tree = {
+                            {
+                                retrieve = "all",
+                                node_type = "parameter_list",
+                                subtree = {
+                                    { retrieve = "all", node_type = "identifier", extract = true },
+                                    {
+                                        retrieve = "all",
+                                        node_type = "typed_parameter",
+                                        subtree = { { retrieve = "all", node_type = "identifier", extract = true } },
```
Comment (danymat) [inline lua/neogen/configurations/julia.lua:35]: Maybe try with 

```lua
subtree = {
  {
    position = 1,
    extract = true,
    as = I.Identifier
  },
    {
    position = 2,
    extract = true,
    as = ...
  },
}
```

Comment (danymat) [reply lua/neogen/configurations/julia.lua:35]: if this doesn't work, just retrieve the parent node aka `typed_parameter` and do a Lua separation with `::`.

File: lua/neogen/configurations/julia.lua [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                            --     node_type = "type",
+                            --     as = i.ReturnTypeHint,
+                            --     extract = true,
+                            -- },
+                        }
+                        local nodes = nodes_utils:matching_nodes_from(node, tree)
+                        if nodes["typed_parameter"] then
+                            results["typed_parameters"] = {}
+                            for _, n in pairs(nodes["typed_parameter"]) do
+                                local type_subtree = {
```
Comment (danymat) [inline lua/neogen/configurations/julia.lua:86]: look at comment above

File: lua/neogen/configurations/julia.lua [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                                end
+                            end
+                        end
+                    end
+                    return 0, 0
+                end
+            end,
+        })
+        :add_default_annotation("google_docstrings")
+        :add_annotation("numpydoc")
```
Comment (danymat) [inline lua/neogen/configurations/julia.lua:169]: Be sure to create your own Julia convention when it progresses

**Comment by josephsdavid** [comment]
@danymat I believe this is good to be merged! As far as conventions go, I am still trying to figure out if/where any official guidelines exist, currently I am just following all the documentation examples I have seen so far!

**Comment by josephsdavid** [comment]
actually hold on that, after reading https://docs.julialang.org/en/v1/manual/documentation/ more thoroughly i have some changes

**Comment by josephsdavid** [comment]
I think for now this is probably sufficient but I will need to add more as I understand how julia documentation works!

**Comment by danymat** [comment]
Hello ! Very cool ! Can you share me some code examples of your features, and tell me which convention you used ?

**Comment by danymat** [comment]
Hello @josephsdavid , what's the state of the integration ?

**Comment by josephsdavid** [comment]
Oh! I have forgotten to finish the job here, let me make these things!

**Comment by josephsdavid** [comment]
So as far as conventions, Julia does not have established conventions like numpydoc jsdoc emmylua etc, so I guess we could just call it something like "Julia Alternative" or "Julia unspecified", it is based off of how we document open source stuff at my job, [example](https://beacon-biosignals.github.io/Lighthouse.jl/dev/#Lighthouse.evaluation_metrics_row), so I guess we could call it "Beacon format" for lack of a better name!


```julia
import Base: +, *, /

"""
    Point

A type representing a point in 2d space

Where...

    x::Number is the x coordinate
    y::Number is the y coordinate
"""
struct Point
    x::Number
    y::Number
end

# Open question: Generate docs for this type of function
+(a::Point, b::Point) =  Point(a.x + b.x, a.y + b.y)

+(a::Point, b::Number) = Point(a.x + b, a.y + b)

*(a::Point, b::Point) = Point(a.x * b.x, a.y * b.y)

/(a::Point, b::Number) = Point(a.x / b, a.y / b)



"""
    f(x,y,z,w)

f does some operations on some points and numbers

Where...

    z::Number is a shift factor, which shifts a point
    w::Number is a scaling factor, which scales the output
    x is the first point
    y is the second point
"""
function f(x, y, z::Number = 1, w::Number = 3)
    return (x * (y + z)) / w
end
```
above is a little code demo, two sort of open questions are:
1. how to document the f(x) = 2 * x syntax
2. Need to be able to grab the function title and arguments, currently this is [done manually](https://sendvid.com/tt7b76dp)

**Comment by f3fora** [comment]
Actually Julia defines some conventions in the [documentation](https://docs.julialang.org/en/v1/manual/documentation/) or more accurately in [Documenter.jl](https://juliadocs.github.io/Documenter.jl/stable/).

Arguments are actually introduced with:

```
"""
...
# Arguments
- `n::Integer`: the number of elements to compute.
- `dim::Integer=1`: the dimensions along which to perform the computation.
...
"""
```

**Comment by josephsdavid** [comment]
> Actually Julia defines some conventions in the [documentation](https://docs.julialang.org/en/v1/manual/documentation/) or more accurately in [Documenter.jl](https://juliadocs.github.io/Documenter.jl/stable/).
> 
> Arguments are actually introduced with:
> 
> ```
> """
> ...
> # Arguments
> - `n::Integer`: the number of elements to compute.
> - `dim::Integer=1`: the dimensions along which to perform the computation.
> ...
> """
> ```

Ah let me change this!

**Comment by f3fora** [comment]
the format is currently wrong ("- `\t%s::%s` is $1")
It should be "- `%s::%s` : $1"

**Comment by josephsdavid** [comment]
new code sample! 

```julia
import Base: +, *, /

"""
    Point(x,y)

A struct representing a 2d point

# Arguments

- `x::Number` is the x coordinate
- `y::Number` is the y coordinate
"""
struct Point
    x::Number
    y::Number
end

# Open question: Generate docs for this type of function
+(a::Point, b::Point) =  Point(a.x + b.x, a.y + b.y)

+(a::Point, b::Number) = Point(a.x + b, a.y + b)

*(a::Point, b::Point) = Point(a.x * b.x, a.y * b.y)

/(a::Point, b::Number) = Point(a.x / b, a.y / b)



"""
    f(x,y,z,w)

f does some operations on some points and numbers

# Arguments

- `z::Number` is a shift factor, which shifts a point
- `w::Number` is a scaling factor, which scales the output
- `x` is the first point
- `y` is the second point
"""
function f(x, y, z::Number = 1, w::Number = 3)
    return (x * (y + z)) / w
end

a = Point(1,2)
b = Point(1,3)

f(a,b)
```

**Comment by f3fora** [comment]
According to the documentation:
- in [struct](https://docs.julialang.org/en/v1/manual/documentation/#Types) each field can be documented independently with no need of adding `arguments` in the struct docstring.
- in [arguments](https://docs.julialang.org/en/v1/manual/documentation/#Functions-and-Methods) there is a `:` instead of `is`.

**Comment by josephsdavid** [comment]
> According to the documentation:
> 
> * in [struct](https://docs.julialang.org/en/v1/manual/documentation/#Types) each field can be documented independently with no need of adding `arguments` in the struct docstring.
> * in [arguments](https://docs.julialang.org/en/v1/manual/documentation/#Functions-and-Methods) there is a `:` instead of `is`.
> 
> Moreover, the default argument value should be included and the arguments should be in order.

Thanks! I will knock this out after work :)

**Comment by josephsdavid** [comment]
So need to figure out how to get the arguments in order, it is tricky because I have to parse typed and nontyped arguments separately, and default arguments I need to figure out as well (and typed default arguments 🤯 !) So this will likely turn into a weekend project!

**Comment by danymat** [comment]
> So need to figure out how to get the arguments in order, it is tricky because I have to parse typed and nontyped arguments separately, and default arguments I need to figure out as well

I'm afraid that's a limitation at the moment..
What's the state of the PR ?

**Comment by josephsdavid** [comment]
> > So need to figure out how to get the arguments in order, it is tricky because I have to parse typed and nontyped arguments separately, and default arguments I need to figure out as well
> 
> I'm afraid that's a limitation at the moment.. What's the state of the PR ?

if thats the limitation, it seems i just need to get the default arguments to show up and then it is good!

**Comment by josephsdavid** [comment]
> > > So need to figure out how to get the arguments in order, it is tricky because I have to parse typed and nontyped arguments separately, and default arguments I need to figure out as well
> > 
> > 
> > I'm afraid that's a limitation at the moment.. What's the state of the PR ?
> 
> if thats the limitation, it seems i just need to get the default arguments to show up and then it is good!

Still stuck here lol but have not forgotten

**Comment by danymat** [comment]
Hello, what's the current state of this PR ? What are the current things I can catch on and try to merge it ?

**Comment by danymat** [comment]
Hello @josephsdavid, I can finish implementing this PR if you think it's in a good progress. If so, could you pinpoint me what needs to be continued ?

**Comment by danymat** [comment]
Hello @josephsdavid , #185 is being merged for julia support on neogen ! I will close this issue, now that the other one can be marked as the new reference.

Thanks again for all your preliminary help. Best,
Dany.

**Pull request closed without merging**